### ISSUE TITLE
default storage class for RWO vs RWM

### ISSUE THREAD

**Issue by kfox1111** [kubernetes/kubernetes#89911]
We have multiple storage classes and its always unclear what to set as default, as there can be one best default for ReadWriteOnce vs ReadWriteMany.

I'd like the option to mark one storage class that is ReadWriteOnce as storageclass.kubernetes.io/is-default-class and one that is ReadWriteMany as storageclass.kubernetes.io/is-default-class and have Kubernetes pick the right default based on the users ReadWriteX request in the pvc.

**Comment by 4censord** [comment]
😭

**Comment by kfox1111** [comment]
Once cel based mutating policies are a thing, that may be an easy way to support it external from k8s.

**Comment by sunnylovestiramisu** [comment]
The current behavior is:

"If you set the [storageclass.kubernetes.io/is-default-class](https://kubernetes.io/docs/reference/labels-annotations-taints/#storageclass-kubernetes-io-is-default-class) annotation to true on more than one StorageClass in your cluster, and you then create a PersistentVolumeClaim with no storageClassName set, Kubernetes uses the most recently created default StorageClass."

https://kubernetes.io/docs/concepts/storage/storage-classes/

**Comment by AndrewSirenko** [comment]
I've heard of multiple AWS EKS customers who wanted something like this. E.g. they wanted to rely on EBS for their RWO volumes, but EFS/FSx for RWM,ROM volumes, without explicitly specifying an SC for each workload.  

But question 3 points out a tricky UX.

> 1. if the IsRequestedAccessModeAnnotation is missing, should we fall back to the default StorageClass without accessMode specified?

Makes sense. This is backwards compatible. CSI driver of default SC can fail volume creation with InvalidArgument if it doesn't support PVC's combination of SC + AccessMode(s).

> 2. Do we need to consider accessMode in quota counting as well?

Sounds out of scope for this issue/feature to me? Even if user has use-case like "SC Gold should only be used for RWO volumes," I don't think whether the SC is default or not has implications for that quota?

> 3. If there are multiple accessModes set, which default should we apply? For example do we pick the first accessMode in the list?

Because SCs have differing cost, performance, and pod scheduling flexibility implications, if cluster admin does not consider PVCs with multiple access modes, they may surprise developers by adding this accessMode annotation to default SCs.

Maybe it's safest to have cluster admin explicitly set each default SC's combination of access modes like:  `IsRequestedAccessModeAnnotation=RWO,[RWO,RWM]`

And if:
1. All default storage classes have an accessModeAnnotation,
2. Some combination of access modes is missing

Then we'll fail the volume binding and leave event on PVC.

Cluster admin can avoid this inconvenience by setting a default SC without accessMode specified to fallback to old behavior.

**Comment by xing-yang** [comment]
> Questions:

> 1. if the IsRequestedAccessModeAnnotation is missing, should we fall back to the default StorageClass without accessMode specified?
Yes. We need to support backward compatibility.

> 2. Do we need to consider accessMode in quota counting as well?
If we keep the existing quota counting by StorageClass, that should be sufficient?  Since we need to support backward compatibility, we can't always do quota counting by accessMode.

> 3. If there are multiple accessModes set, which default should we apply? For example do we pick the first accessMode in the list?
That's a tricky one. We could either pick the first accessMode or assign the StorageClass to all the accessModes?  This one needs some discussions.

**Comment by sunnylovestiramisu** [comment]
For 2, we discussed that for StorageClass, we want to add it to the quota Scope eventually. It will be a brand new way to specify Quota. Maybe there is a way to implement accessMode support.

**Comment by kfox1111** [comment]
How is quota counted today? if it is defaulted, does the volume count against the assigned storageclass name, or nothing? I would expect it to be counted under the assigned storageclass, but I guess I've never tested that.

**Comment by AndrewSirenko** [comment]
Thanks for re-opening this, many EKS customers have wanted by default to rely on EBS volumes for RWO, and EFS for RWM ROM.

Re 1: Makes sense for backwards compatibility.

Re 2: Sounds out of scope for this issue/feature to me? Even if user has use-case like "SC Gold should only be used for RWO volumes," I don't think whether the SC is default or not has implications for that quota? 

Re 3: Agreed with Jan's suggested plan. Looks like [csi-lib-utils](https://github.com/kubernetes-csi/csi-lib-utils/blob/e0fa89db2e21091a8d0ce49609ebef3ff8533b58/accessmodes/access_modes.go#L47) also gives precedence to RWX (but eventually [kubelet just picks first in list at NodeStage/Publish time](https://github.com/search?q=repo:kubernetes/kubernetes+accessmodes%5B&type=code).

**Comment by thockin** [comment]
A few questions:

1) I have seen read-write-many shortened to "rwx" and "rwm" -- which is correct?

2) I don't love having multiple annotations - it is putting the data (mode) into the name.  `storageclass.kubernetes.io/is-default-class-for-mode: rwm,rwo,rom` might be cleaner?

3) Is this a place where a MutatingAdmissionPolicy would be better?  That would (IIUC) have access to the body of the PVC, so could look at `.spec.volumeMode` + `.spec.storageClassName` and make a decision about which class to set.  It disconnects the choice of class from the class object itself, but it centralizes the logic for which class to use (and offers even more flexibility).

@cici37  @jpbetz as the purveyors of `*AdmissionPolicy` APIs - opinion?

**Comment by kfox1111** [comment]
I like storageclass.kubernetes.io/is-default-class-for-mode: rwm,rwo,rom as an api

There is ReadWriteOncePod too, but I think we can just use rwo to cover that? I dont think there really is a use case where those might need to be different for a default that I can think of.

Having written a MutatingAdmissionPolicy, I think it could be done to implement this. Though I think for the vast majority of clusters, it would always look the same. one storage class for rwm, one for rwo. If we think MutatingAdmissionPolicy should replace this use case, we should probably deprecate the old default storage class admission stuff too and then remove it. A MutatingAdmissionPolicy that forces a default is even easier with MutatingAdmissionPolicy then the RWO vs RWM logic.

Its a common enough use case though, I can see the value of supporting it directly in the apiserver though. Most users may want to do some form of it and then dont have to learn cel to do so.

**Comment by thockin** [comment]
I understand the idea that "built-in" seems realer, but I also want to try to balance how much we build IN versus how much we build ON kubernetes.

We have thus far avoided This one seems pretty safe in isolation, but a lot of these end up as very slippery slopes.  I would not, for example, want to see a whole family of policy oriented annotations which govern defaulting.  Like, if you had a different default per namespace or per tenant?  That is exactly what MAP is for, so why not this? 

I'm not saying I'm against it, per se, but I want to understand what principles would say yes to this but no to a per namespace setting.  I mean, annotations are arbitrary, so someone could write an admission controller which read Lua code from an annotation on the class and interprets it against claims. I think that would be nuts, but it's totally valid.

We have, thus far, avoided adding storage policy APIs as resources.

**Comment by kfox1111** [comment]
Yeah. I get that. I almost closed this, almost 4 year old issue when I saw it could possibly be done with MutatingAdmissionPolicy. (Still a while before map is listed as stable...)

But, wanted a little discussion around it first as discussion has started up.

I do think your right about it being a slippery slope, and that's why I also brought up deprecating the default logic entirely.

The value is around the tradeoff on:
What proportion of all admins will use the feature, out of the box on their cluster. If its 90% or more, its probably worth keeping in the codebase rather then out of it? I could see it getting that far.

How many of the existing users today would use the RWO vs RWM feature over the single default if it exists? I think a lot. I would. But that's a gut feeling.

The admins today have a catch 22.... They can set a default, but it will be wrong for a good portion of workloads and will have to field user support issues on why things are not working well ("My second pod wont start", "Its because it can only be attached to one at a time. now you need to play a fancy game of copying data between volumes"). Or, they can force the user to pick, by not setting a default, and then have to field a different set of issues of a different form ("I created a pvc but its stuck pending". "You need to delete the pvc and specify the storage class next time. If its a statefulset you have to delete that too.")

Either way, the admins are fielding issues related to this sharp edge. Its toil. :/

I don't set defaults on mine today because fixing the wrong default is worse then fixing the stuck pvc related issues. But not every admin realizes that until it has bit them a few times and maybe to late to fix the default.

Having it in tree, and in docs on the right way to do it would be really nice.. But recommending the current, default without consideration to type is painful. :/

I consider the current implementation kind of dangerous...

So, that leads me back to my recommendation. remove it entirely and go the MutatingAdmissinPolicy route totally, or add enough support to the default code to make it not dangerous? I have a bit of a preference to it being in tree, but could totally live with deprecation/map.

**Comment by sunnylovestiramisu** [comment]
If we use storageclass.kubernetes.io/is-default-class-for-mode key value pair, the value can stay the same as defined in access modes without being shortened, which is ReadWriteOnce, ReadOnlyMany, ReadWriteMany, ReadWriteOncePod.

@jsafrane is suggesting we introduce an order of these modes: RWM > RWO > RWOP and RWM > ROM because:

- If there is a SC that can provision say RWO, it probably supports RWOP (it's just a special case of RWO to me). Do we know that it supports RWM? We don't.
- If there is a SC that can provision RWM, it can probably support RWO (just mount the volume to a single node) and RWOP.
- if something can provision RWM, then ROM is just RWM mounted as read-only.

However, we cannot compare ROM and RWO/RWOP.

**Comment by kfox1111** [comment]
I wouldn't assume RWM can support RWO. usually RWM volumes are things like nfs or nfs like filesystems (cephfs, lustre, etc) and they use a ... loose definition of posix that isn't always compatible with things like databases.

Thats kind of why we are in this boat right now I think. A single default that is an RWO wont work for services that expect multiple attachment. And a single default of a RWM wont work well for databases or other things that expect sole access to the volume, and need all of posix to function properly.

**Comment by jpbetz** [comment]
> 3. Is this a place where a MutatingAdmissionPolicy would be better?  That would (IIUC) have access to the body of the PVC, so could look at `.spec.volumeMode` + `.spec.storageClassName` and make a decision about which class to set.  It disconnects the choice of class from the class object itself, but it centralizes the logic for which class to use (and offers even more flexibility).
> 
> @cici37 @jpbetz as the purveyors of `*AdmissionPolicy` APIs - opinion?

At a glance this looks like it should be possible using MutatingAdmissionPolicy.

I don't have anywhere near enough context to assess if this best left up to cluster admi8ns to solve via an extension (such as MutatingAdmissionPolicy) or if this is sufficiently core to users needs that a built-in solution is needed/justified.

If anyone tries this out with MutatingAdmissionPolicy and runs into problems, please feel free to reach out. This appears to be a policy problem that should be possible to solve with MutatingAdmissionPolicy.

**Comment by thockin** [comment]
Exploring further, why wouldn't we want something like a `PersistentVolumePolicy` object which has a namespace selector and some explicit default values for classes?  Are there other things that people struggle with?  Default size or granule?  Access control (who can use which classes)?  Default mode?

**Comment by kfox1111** [comment]
Yeah, I didn't mean literally remove. but hard deprecate and point at the better option, whatever it is.

The main problem as I see it is that posix behavior. Is your app expecting multiple attachment, or single ownership. That expected behavior is roughly encoded in access mode.

A PersistentVolumePolicy object would work too over an annotation. It has the nice property of being easier to schema check.

Default size could be interesting, but really kind of hard to guess for the user. Though guessing low is usually better then guessing high. There is some precedent to being able to set a default set by LimitRange.

Being able to namespace select could be nice for some users. Say you have a pool of nodes per tenant, and a set of storage classes per tenant...

**Comment by sunnylovestiramisu** [comment]
The annotation approach needs to go through alpha/beta/GA feature gate, so it will also take about a year.

**Comment by kfox1111** [comment]
Ok, so, we started to look at how to implement this using a MAP.... We ran into something unexpected though..

`AccessModes` in the PVC is an enum...

I've only ever seen it with a single value. How does it behave if multiple are specified?

**Comment by sunnylovestiramisu** [comment]
> Ok, so, we started to look at how to implement this using a MAP.... We ran into something unexpected though..
> 
> `AccessModes` in the PVC is an enum...
> 
> I've only ever seen it with a single value. How does it behave if multiple are specified?

Hmm seems like MAP runs after intree defaultstorageclass admission controller. How did you resolve that?

**Comment by kfox1111** [comment]
I didnt use the intree one.
